### DISCUSSION TITLE
A possibility of some feature that shows voxel number (Nx,Ny,Nz) for specific voxel on mcMatlab plots

### DISCUSSION THREAD

**Discussion by TZ387** [ankrh/MCmatlab#23] [Ideas]
Dear Anders and Dominik,

could some feature that shows the exact voxel number (Nx,Ny,Nz) for specific voxel (f.e., the one that is pointed on by cursor) be added to the code? As far as I am aware, such option currently isnt available, but it could be useful in some cases, such as when we are dealing with very small media.

To roughly illustrate what I mean (of course, this concept can be also achieved in many other ways), here is ad-hoc code that I added to NdimSliderPlot.m (at the end of NdimSliderPlot function) so that the data tip option for matlab graphs shows voxel number in addition to their location:

```
  % Enable data cursor mode
  dcm_obj = datacursormode(gcf);
  set(dcm_obj, 'UpdateFcn', {@customDataTip, xraw, yraw, zraw})

  function output_txt = customDataTip(~, event_obj, xraw, yraw, zraw)
     
      pos = get(event_obj, 'Position');
      x = pos(1); y = pos(2); z = pos(3);

      [~, index_x] = min(abs(xraw - x));      
      [~, index_y] = min(abs(yraw - y));   
      [~, index_z] = min(abs(zraw - z));

      output_txt = {['X: ', num2str(x), ' (', num2str(index_x), ')'], ...
          ['Y: ', num2str(y), ' (', num2str(index_y), ')'], ...
          ['Z: ', num2str(z), ' (', num2str(index_z), ')']};

   end
```

What I get can be seen on image below (from Example4_BloodVeseel.m):

<p align="center">
<img src="https://github.com/ankrh/MCmatlab/assets/135353620/28eaf2d5-97d4-424e-8149-149fbbb14db3" width="480">
</p>

Best regards, Tilen

**Comment by ankrh** [comment]
It's a good suggestion.

But one challenge is that the locations that the data cursor snaps to is always the _corners_ of the voxels, which means that for every xyz position the data cursor shows, there are 8 voxels equally far away from it. I'm not sure what the best way to deal with this is. One possibility is to let the index shown be fractional, such as stating "50.5" as the index that the cursor is at. But this also looks a bit ugly and counterintuitive.

What do you think?

**Reply by TZ387** [reply]
You are right, I did not consider this fact. I will check this problem in the following days to try to find a better
approach. But I do agree that "50.5" approach is unintuitive - In most cases, I guess that I would prefer a whole number
(and perhaps paraphrase it a bit differently - f.e., number of voxels transversed in some direction - from 0 to Nmax, or consecutive coordinate number - from 1 to Nmax + 1).

**Comment by TZ387** [comment]
Now I made a more thorough check-up of the problem (Including checking possible UI alternatives to data tip option, as well as options/properties available for DataCursorManager),  but I am afraid that I didn't find any easy/great solution on my side.

Two options that I see as most sensible are:

1. Rephrase the values obtained with data tips - the same thing that I mentioned in my previous comment. 

2. The second option I see is to add some (invisible) data points located at the center of the voxels to the plots. In that case, center points could be represented by whole number - "50", while the edges could be shown with fractional - "50.5". Alternatively, the plotting code could be changed in such a way that only center of voxels are shown when using data tips.

The option 2 could be interesting in case plot values (fluence, temperature etc.) are added to data tips in the future, but given the fact that this would complicate things and take some effort and time, I guess that for now I would go with option 1 or don't make any changes to code at all

p.s.: When writing the custom update function, I probably should have included the original function. It is

```
function output_txt = myfunction(obj,event_obj)
% Display data cursor position in a data tip
% obj          Currently not used
% event_obj    Handle to event object
% output_txt   Data tip text, returned as a character vector or a cell array of character vectors

pos = event_obj.Position;


%********* Define the content of the data tip here *********%

% Display the x and y values:
output_txt = {['X',formatValue(pos(1),event_obj)],...
    ['Y',formatValue(pos(2),event_obj)]};
%***********************************************************%


% If there is a z value, display it:
if length(pos) > 2
    output_txt{end+1} = ['Z',formatValue(pos(3),event_obj)];
end

%***********************************************************%

function formattedValue = formatValue(value,event_obj)
% If you do not want TeX formatting in the data tip, uncomment the line below.
% event_obj.Interpreter = 'none';
if strcmpi(event_obj.Interpreter,'tex')
    valueFormat = ' \color[rgb]{0 0.6 1}\bf';
    removeValueFormat = '\color[rgb]{.25 .25 .25}\rm';
else
    valueFormat = ': ';
    removeValueFormat = '';
end
formattedValue = [valueFormat num2str(value,4) removeValueFormat];
```

**Reply by ankrh** [reply]
Thanks for giving this more thought. I also can't think of a more elegant solution than fractional indices. So I think for now I won't be adding such a functionality to MCmatlab. But the good thing about open source is that anyone else reading this is welcome to fork the repo and add his/her own displays to the data cursor.
